## Latihan Penanganan Missing Value dg dataset Titanic


import library Python yang diperlukan untuk analisi data: Numpy dan Panda

In [7]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.tree import DecisionTreeClassifier # Import Decision Tree Classifier
from sklearn.model_selection import train_test_split # Import train_test_split function
from sklearn import metrics #Import scikit-learn metrics module for accuracy calculation


### Load data dan memahami data

Untuk secara detil analisis eksplorasi data secara singkat, dapat melihat ke Notebook titanic00_EDA


In [8]:
df = pd.read_csv('../dataset/titanic/train.csv')
dft = df.copy()


### Teknik 1. Penanganan _Missing value_ dengan **identifikasi**
Mengganti _missing value_ dengan nilai tertentu seperti NA atau Unknown.

Berdasarkan hasil analisi data di sebelumnya, dapat diketahui bahwa: 
* field/kolom/atribut **PassengerId** karena hanya sebagai nomor urut atau unique variabel untuk setiap record, tidak kita perlukan. 
* field/kolom/atribut **Name** memiliki jumlah **unique** sebanyak jumlah record total, yaitu 891, sehingga tidak signifikan untuk pemodelan prediksi.
* field/kolom/atribut **Ticket** memiliki jumlah **unique** hampir sebanyak jumlah record total, yaitu 681 (85%), sehingga tidak signifikan untuk pemodelan prediksi.
* ada 3 field yang memiliki _missing value_ yaitu: age, cabin dan embarked. 

Kemudian untuk record dengan _missing value_ yang akan kita gantikan dengan NA adalah atribute sebagai berikut:
**Cabin, Age dan Embarked**.

In [9]:
dft.drop(["PassengerId","Name","Ticket"], axis=1, inplace=True)

In [10]:
dft.isnull().sum()

Survived      0
Pclass        0
Sex           0
Age         177
SibSp         0
Parch         0
Fare          0
Cabin       687
Embarked      2
dtype: int64

### Sekali lagi, **Eliminasi** untuk record yang memiliki _missing value_
Berdasarkan data di atas, masih ada 2 atribut yang memiliki _missing value_ Age = 177 record dan Embarked = 2 record. 

Langkah berikut adalah menghapus record-record yang field Age atau Embarked memiliki _missing value_. 

In [6]:
dft.drop(["Age","Embarked"], axis=1, inplace=True)
dft.isnull().sum()

Survived    0
Pclass      0
Sex         0
SibSp       0
Parch       0
Fare        0
dtype: int64

In [7]:
dft.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   SibSp     891 non-null    int64  
 4   Parch     891 non-null    int64  
 5   Fare      891 non-null    float64
dtypes: float64(1), int64(4), object(1)
memory usage: 41.9+ KB


### Pemodelan dengan data dengan algoritma Decision Tree

#### Seleksi Fitur (_Feature Selection_)
Sebagaimana algoritma _supervised_ lainnya, kita perlu membagi atribut menjadi 2 bagian. Satu bagian terdiri dari 1 atribut, sebagai variabel _**target**_, dan sisanya adalah sebagai variabel _**independent**_. 

Dalam hal ini, atribute **Survived** adalah variabel target, dan sisanya adalah variabel independent. Kemudian, kita membagi data menjadi data training (learning) dan data testing. Kali ini kita memilih cara pembagian data yang sederhana yaitu data testing adalah 30% dari total dataset.

Selanjutnya, untuk pemodelan dengan algoritma Decision Tree, jenis data harus numerik. Sedangkan kondisi dataset sekarang masih ada beberapa yang non-numerik,  yaitu Sex:male,female. Oleh karena itu atribute Sex harus ditransformasi menjadi nilai numerik. Dengan menggunakan tool LabelEncoder dari sklearn.preprocessing, kita dapat merubah secara otomatis atribute Sex menjadi male = 1, dan female = 0. 


In [14]:
le = LabelEncoder()
dft['Sex'] =  le.fit_transform(dft['Sex'])

In [15]:
dft.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    int64  
 3   SibSp     891 non-null    int64  
 4   Parch     891 non-null    int64  
 5   Fare      891 non-null    float64
dtypes: float64(1), int64(5)
memory usage: 41.9 KB


In [28]:
#split dataset in features and target variable
feature_cols = ['Pclass','Sex','SibSp','Parch','Fare']
X = dft[feature_cols] # Features
y = dft.Survived # Target variable

In [29]:
# Split dataset into training set and test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) # 70% training and 30% test

#### Training: membangun model prediksi dengan Decision Tree

In [30]:
# Create Decision Tree classifer object
clf = DecisionTreeClassifier()

# Train Decision Tree Classifer
clf = clf.fit(X_train,y_train)

#Predict the response for test dataset
y_pred = clf.predict(X_test)

#### Evaluasi akurasi dari model yang dihasilkan: 


In [31]:
# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.7574626865671642
